api/v1beta1/nutanixcluster_types.go

```diff
@@ -152,13 +150,13 @@ func (ncl *NutanixCluster) SetConditions(conditions capiv1.Conditions) {
 func (ncl *NutanixCluster) GetPrismCentralCredentialRef() (*credentialTypes.NutanixCredentialReference, error) {
 	prismCentralInfo := ncl.Spec.PrismCentral
 	if prismCentralInfo == nil {
-		return nil, nil
+		return nil, fmt.Errorf("prismCentral info is not provided.")
 	}
 	if prismCentralInfo.CredentialRef == nil {
-		return nil, fmt.Errorf("credentialRef must be set on prismCentral attribute for cluster %s in namespace %s", ncl.Name, ncl.Namespace)
+		return nil, nil // returning nil so that we can use default controller's secret
```

Comment:
The original purpose was that if prismCentralInfo is not present, prismCentralInfo including credentials of the prism central on which management cluster is running will be used. If it is provided, the credentials were mandatory because you can't just expect credentials of one prism central to work on another. The switchover of these two conditions with that context is rather confusing to me and the motivation for doing so unclear.

Reply:
ok but per public docs, thats what is expected https://opendocs.nutanix.com/capx/v1.3.x/credential_management/. if you dont provide then use creds of controller and if given then use overridden ones. 
also per type def https://github.com/nutanix-cloud-native/prism-go-client/blob/main/environment/credentials/types.go#L130 its specified as optional as well

we can debate separately on the confusing part of this from user perspective but this is the designed behavior at this time in my understanding. cc @yannickstruyf3

Reply:
IMO, in that case, it's the documentation that should be fixed not the code to match unclear documentation. My understanding of what this change was originally supposed to be was to make credentials always mandatory and do away with optional credentials behavior. This PR is no longer doing what the original was doing.

Reply:
We have following cases:

Controller manager uses PC 1 with creds for user1/pwd1
- Workload Cluster1 need to be created on same pc PC 1 with same user as controller used
1. One option: Here technically entire prismCentral section can be optional and we use endpoint and secret used by controller
2. 2nd option (current implementation matching docs with mandatory prismCentral section fix): User specifies prismCentral with same endpoint and keeps CredentialRef empty assuming we use same secret as controller.

- Workload Cluster 2 needs to be created on same pc PC 1 with different user as controller used
Here prismCentral section is required with same pc endpoint but different user creds secret referred thru CredentialRef

- Workload Cluster 3 needs to be created on PC 2 with different user from that PC
Here prismCentral section is required with pc2 endpoint and different user from that PC

Are you proposing that we should not support bullet 1, option 2 and make option 1 available?

Reply:
For the sake of this comment, I'll refer to the three unnamed bullet cases above as a, b, and c (in the order they appear). 

a. Current implementation entails either 1 or 3 can be used to satisfy this requirement.
b. Current implementation entails only 3 can be used to satisfy this requirement.
c. Current implementation entails only 3 can be used to satisfy this requirement.

To be more explicit, 2 has never been a supported mechanism. To that extent, we even have explicit E2E test cases to ensure new changes don't accidentally change the behavior to support 2 and remove 1. 
* https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/blob/74c5d337fe852b904b628d3d19a74f0949c37b1c/test/e2e/nutanix_client_test.go#L75-L129 ensures 2 is explicitly not supported
* https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/blob/74c5d337fe852b904b628d3d19a74f0949c37b1c/test/e2e/nutanix_client_test.go#L134-L176 ensures 1 is explicitly supported

Now for the proposal of removing support for 1. I was under the impression that this was originally your proposal. This is what the original PR https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pull/400 was doing. It entailed removing the optionality of relying on "defaults" altogether. I highlighted this in a comment on the original PR https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pull/400#discussion_r1558250883. It was a breaking change but that was something I could get behind as it simplifies the controller by ensuring credentials and prism central are always required for each cluster.

I am advocating for one of these two (using alpha (ɑ) and beta (β) as we're now running out of bullet indexes):
ɑ. leave things as is: i.e. prism central info stays optional; if it's left blank, prism central info and credentials are taken from default CAPX config. If the prism central info is specified, credentials have to be explicitly provided. 
β. make both mandatory: no more clusters relying on default prism central and credentials i.e. remove support for 1.

Reply:
The first comment made by @thunderboltsid explains the initial intent of the code. The documentation can be modified to reflect the same.